src/PointwiseFunctions/AnalyticData/GrMhd/CocalInitialData.hpp

```diff
@@ -0,0 +1,216 @@
+        });
+    // Compute derived quantities from interpolated data
+    // const size_t num_points = x.begin()->size();
+    // const auto& rest_mass_density =
+    //     get<hydro::Tags::RestMassDensity<DataVector>>(result);
+    const auto& specific_internal_energy =
+        get<hydro::Tags::SpecificInternalEnergy<DataVector>>(result);
+    const auto& spatial_velocity =
+        get<hydro::Tags::SpatialVelocity<DataVector, 3>>(result);
+    const auto& pressure = get<hydro::Tags::Pressure<DataVector>>(result);
+    const auto& spatial_metric =
+        get<gr::Tags::SpatialMetric<DataVector, 3>>(result);
+    // Compute enthalpy from internal energy and pressure
+    // auto& specific_enthalpy =
+    //     get<hydro::Tags::SpecificEnthalpy<DataVector>>(result);
+    // get(specific_enthalpy) = DataVector(num_points);
+    // for (size_t i = 0; i < num_points; ++i) {
+    //   const double local_rest_mass_density = get(rest_mass_density)[i];
+    //   if (equal_within_roundoff(local_rest_mass_density, 0.)) {
+    //     get(specific_enthalpy)[i] = 1.;
+    //   } else {
+    //     get(specific_enthalpy)[i] =
+    //     get(hydro::relativistic_specific_enthalpy(
+    //         Scalar<double>(local_rest_mass_density),
+    //         Scalar<double>(get(specific_internal_energy)[i]),
+    //         Scalar<double>(get(pressure)[i])));
+
+    // Constant electron fraction specified by input file
+    auto& electron_fraction =
+        get<hydro::Tags::ElectronFraction<DataVector>>(result);
+    get(electron_fraction) = DataVector(num_points, electron_fraction_);
+    // Zero magnetic field and divergence cleaning field
+    auto& magnetic_field =
+        get<hydro::Tags::MagneticField<DataVector, 3>>(result);
+    get<0>(magnetic_field) = DataVector(num_points, 0.);
+    get<1>(magnetic_field) = DataVector(num_points, 0.);
+    get<2>(magnetic_field) = DataVector(num_points, 0.);
+    auto& div_cleaning_field =
+        get<hydro::Tags::DivergenceCleaningField<DataVector>>(result);
+    get(div_cleaning_field) = DataVector(num_points, 0.);
+    // Compute Lorentz factor from spatial velocity
+    auto& lorentz_factor = get<hydro::Tags::LorentzFactor<DataVector>>(result);
+
+    get(lorentz_factor) =
+        1. / sqrt(1. - get(dot_product(spatial_velocity, spatial_velocity,
+                                       spatial_metric)));
+
+
+    // Set temperature to zero for now
+    auto& temperature = get<hydro::Tags::Temperature<DataVector>>(result);
+    get(temperature) = DataVector(num_points, 0.);
+    return result;
+  }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& /*p*/) override;
+
+ private:
+  std::string data_directory_{};
+  CocalIdType id_type_;
+  double electron_fraction_ = std::numeric_limits<double>::signaling_NaN();
+
+  // This lock is used to ensure that only one thread at a time is calling the
+  // COCAL interpolation routines. We make some assumptions here to guarantee
+  // thread-safety:
```

Comment:
assumptions won't guarantee anything. Are these assumptions actually true? If not then this needs to be checked at runtime (and the assumptions *documented* like they are done here).